### PR TITLE
THRIFT-5034 Allow for Reverse Connection servers

### PR DESCRIPTION
**Pull request by ptsneves**
The changes at this point are very shallow as they only include a reference implementation for Python. I will in the course of the discussion add tests and some reference implementations in other languages.

EDIT: The initial commit was wrong and included other things i am working on. Disregard them and only regard the single patch with Jira ID number

### CODE REVIEW

File: lib/py/src/transport/TReverseTunnelServer.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import queue
+from .TTransport import TTransportBase, TTransportException, TServerTransportBase
+
+
+
+class TReverseTunnelServerFactory(object):
+    """Factory transport that builds framed transports"""
```
Comment (Jens-G) [inline lib/py/src/transport/TReverseTunnelServer.py]: comment needs a fix, I guess

File: lib/py/src/transport/TReverseTunnelServer.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import queue
```
Comment (Jens-G) [inline lib/py/src/transport/TReverseTunnelServer.py]: ASF header again

Comment (Jens-G) [reply lib/py/src/transport/TReverseTunnelServer.py]: .

**Comment by ptsneves** [comment]
From what i see the travis-ci is failing only on some c++ thing and these changes are unrelated to c++.

**Comment by RandyAbernethy** [comment]
The Travis complaints are style based but related to this patch:

`# Python code style
flake8
./test/py/TestReverseTunnel.py:35:1: E302 expected 2 blank lines, found 1
./test/py/TestReverseTunnel.py:37:9: F401 'glob' imported but unused
./test/py/TestReverseTunnel.py:57:13: E265 block comment should start with '# '
./test/py/TestReverseTunnel.py:58:13: E265 block comment should start with '# '
./test/py/TestReverseTunnel.py:92:1: E305 expected 2 blank lines after class or function definition, found 1
./lib/py/src/server/TReverseTunnelServer.py:22:1: E265 block comment should start with '# '
./lib/py/src/server/TReverseTunnelServer.py:24:1: E302 expected 2 blank lines, found 1
./lib/py/src/server/TReverseTunnelServer.py:67:9: E265 block comment should start with '# '
./lib/py/src/server/TReverseTunnelServer.py:71:1: E302 expected 2 blank lines, found 1
The command "build/docker/run.sh" exited with 1.`

Simple to fix so probably worth fixing.

File: lib/py/src/server/TReverseTunnelServer.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+from six.moves import queue
+from thrift.transport import TTransport
+#from thrift.transport.TTransport import TTransportBase, TTransportException, TServerTransportBase
```
Comment (dcelasun) [inline lib/py/src/server/TReverseTunnelServer.py]: Don't leave commented imports.

File: lib/py/src/server/TReverseTunnelServer.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+class TReverseTunnelServer(TTransport.TTransportBase):
+    """A transport that allows RPC Servers use Client Transports.
+
+    This useful for reverse connections, where
```
Comment (dcelasun) [inline lib/py/src/server/TReverseTunnelServer.py]: Grammar and structure fixes to this whole section:

```
A transport that allows RPC servers to use client transports.

This is useful for "reverse connections" where a primary server
passes RPC calls to secondary servers attached to it. One example
would be an RPC service in a private network that doesn't allow
public traffic.

The current implementation of this Transport is compatible with
ThreadedPoolServer and other concurrent servers as it blocks calls
to accept() based on the number of active connections.

The idea is that for a server, a flush means a finished RPC call. This
isn't mentioned in the specification, but is a good heuristic for signalling
the accept() caller.
```

Comment (ptsneves) [reply lib/py/src/server/TReverseTunnelServer.py]: I made some changes but could you be more specific, specially on the structure?

File: lib/py/src/server/TReverseTunnelServer.py (hunk 3 above)
Comment (ptsneves) [reply lib/py/src/server/TReverseTunnelServer.py]: done

File: lib/py/src/server/TReverseTunnelServer.py (hunk 4 above)
Comment (dcelasun) [reply lib/py/src/server/TReverseTunnelServer.py]: I mean my version fixes some typos and grammar issues, and it reads better (to me, at least). If you don't have any objections to it, can you use it as is?

**Comment by dcelasun** [comment]
@ptsneves any interest in contributing this for other languages? I'm still tenatively +1 on this, but I think it's important to have at least two implementations of brand new, out-of-spec features like this.

File: lib/py/src/server/TReverseTunnelServer.py (hunk 4 above)
Comment (ptsneves) [reply lib/py/src/server/TReverseTunnelServer.py]: Ahh of course, i thought it was a quote not a suggestion to use that.

**Comment by ptsneves** [comment]
@dcelasun Yes, I will try to submit it. I can do it in C++ or C#. Do you want cross language tests also? If so I do not know how to do this, but if you provide me with a hint i can do it.

**Comment by dcelasun** [comment]
C++ or C# sounds fine, whichever you are more comfortable with.  Cross tests would be great! They are defined [here](https://github.com/apache/thrift/blob/0ebf3ce99d74bc77912536dc3ad95dac10ebf12a/test/tests.json). For languages not covered by your test, you'll have to add known failures [here](https://github.com/apache/thrift/blob/0ebf3ce99d74bc77912536dc3ad95dac10ebf12a/test/known_failures_Linux.json). See a recent example [here](https://github.com/apache/thrift/commit/f1c53418a794f3d740b809ece822ce79c0bb62a3).

If you have any questions or need help, feel free to join our [Slack channel](https://s.apache.org/slack-invite) #thrift and ping me there.

**Pull request closed without merging**